### PR TITLE
Issue 24: IInteractable interface, Consumable, Collectable scripts added.

### PR DESCRIPTION
**Pull request by mgieger**
Adds a couple of additional consumables using color variations of the "Food1" fish sprite.
Adds the IInteractable interface.
Adds the Food consumable.
Adds the HealthPotion consumable.

ISSUE: The current implementation may not be the best way to handle this issue. The "PlayerHealth" variable of the GameManager class is only used to update the player's health when a state change happens (ie changing floors). I believe a solution is to instead have the pickups modify the Player class' "food" variable. Currently implemented in the player class there is a "food" and "soda" tag functionality which should be replaced with the "Pickup" (or interactable) tag and then call the "Interact" function of the IInteract interface.

### CODE REVIEW

File: Assets/Scripts/Consumables/Fish.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Fish : Consumable {
```
Comment (Zonr0) [inline Assets/Scripts/Consumables/Fish.cs]: Do we need a fish class? Can we create a generic food class and then set the health/mana/etc. gain for each on an individual prefab basis?

**Comment by mgieger** [comment]
A generic Food class would probably be a good re-phrasing there. We
discussed this a bit today when we broke into groups and we came up with
two main ideas for how to implement the consumable pickups.
1. Write a class for each general type of pickup which will contain only
variable fields that each specific item/type of item modifies.
2. Write a super-broad pickup class which would contain all variable fields
that all pickups could possibly modify and then for each pre-fab only place
values in the unity editor for those which the specific pickup would modify.

We went with option 1, and so this commit represents how my interpretation
of this. Does this make sense? As you say though, at the very least Fish
should be changed to say Food. Good call.

**Comment by Zonr0** [comment]
Number 1 sounds good to me. It just seems like we can roll any pickup that directly modifies a stat like hp, mp, xp, or any similar volatile value into one class.

**Comment by mgieger** [comment]
I'm all for it.

**Comment by connorkuehl** [comment]
I think both of these strategies have their merits. I think we should do the first one. 

The goal here is flexibility. We want to make it super simple to add an item (health potion, pick-up, money) to the game (for our sakes). We want to write generalized code here that is applicable to most occasions, and that facilitates the easy additions of items to our game.

> 1. Write a class for each general type of pickup which will contain only variable fields that each specific item/type of item modifies. 

This is a reasonable strategy. For example, we could implement it in such a way that `Consumable` will immediately modify whatever Player stat it corresponds to and our `Pick-up` class's `Interact()` might contain a call to something like `Player.AddToInventory()`. 

This way our inheritance hierarchies are not deep, and it still gives us that flexibility we want. It also makes it easy to add items of type `Consumable` (fish, mana water pocket), or `Pick-up` (hats, money, potions). It just becomes a question of creating the prefab, tacking on the `Consumable` script, modify the value we want to target `"health"`, `"money"`, etc, whatever, and boom.

We could even avoid much of an inheritance hierarchy if we were to implement some sort of `Player` interaction API. Note, providing this API is outside of the scope for this issue. For example,

The Player/GameManager could provide functions like `ModifyPlayerStat(String stat, int modifier)` where we could pass in agreed upon String values or enums to target the stat we would like to modify and the modifier. The GameManager or Player (whoever is providing this API), would then parse our string or enum, and update the stat accordingly. 

Similarly, adding money to the Player's inventory would be as simple as an API call to `AddToPlayerInventory(Item toAdd, int quantity)`. Again, these two systems and the API I described as example would require an entire design run down anyway. 

**The point I'm trying to make is that the first option leaves several doors open and it provides an extensible framework for a loot system.**

> 2. Write a super-broad pickup class which would contain all variable fields that all pickups could possibly modify and then for each pre-fab only place values in the unity editor for those which the specific pickup would modify.

This would also give us extreme flexibility, but after typing out some of my initial thoughts on the first one I think it might be better to try the first option. This would be fine on a small scale, but the truth is we don't know how deep our loot tables will go or how many different items the player will be interacting with on a frequent basis. I think the first option gives us something to stand on and it's extensible as is. If we stuck with this, we could potentially be modifying dozens of fields on a per-item basis. Creating the `Consumable` and `Pickup` types of abstractions lend themselves much better to coding and identifying so it's easier for us to handle those _transactions_ between the Player and whatever it is the Player is working with. Not to mention, it will make creating these item types much cleaner, I think.

**Comment by peschkaj** [comment]
Looks good to me. :+1: to merge

**Comment by brancraig** [comment]
For the reasons that Connor stated, I think the first strategy is the better choice.

**Pull request merged**